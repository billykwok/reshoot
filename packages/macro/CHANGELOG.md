# @reshoot/macro

## 2.1.4

### Patch Changes

- ec76336: fix: improve rollup setup and update dependencies
- Updated dependencies [ec76336]
  - @reshoot/loader@2.1.3
  - @reshoot/types@2.0.3

## 2.1.3

### Patch Changes

- 7492d4c: fix: fix rollup exports

## 2.1.2

### Patch Changes

- 66e5ac9: fix: do not use integer as boolean in terser
- Updated dependencies [66e5ac9]
  - @reshoot/loader@2.1.2
  - @reshoot/types@2.0.2

## 2.1.1

### Patch Changes

- 226a185: fix: add node version restriction and fix rollup config
- Updated dependencies [226a185]
  - @reshoot/loader@2.1.1
  - @reshoot/types@2.0.1

## 2.1.0

### Minor Changes

- ebfb354: feat: export ImageMeta abd Options types from packages

### Patch Changes

- Updated dependencies [ebfb354]
  - @reshoot/loader@2.1.0

## 2.0.0

### Major Changes

- 8d162cb: Introducing Reshoot 2

  Reshoot has been rewritten from ground up to support more use cases. Please refer to README.md for details.

### Patch Changes

- Updated dependencies [8d162cb]
  - @reshoot/loader@2.0.0
  - @reshoot/types@2.0.0
