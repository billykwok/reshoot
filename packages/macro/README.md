# <img src="https://raw.githubusercontent.com/billykwok/reshoot/main/logo.png" width="200" alt="Reshoot logo" /><br/>`@reshoot/macro`

A [`babel-macro`](https://github.com/kentcdodds/babel-plugin-macros) transpiling `reshoot('image.png', { color: '#eee' })` into `import meta from 'image.png?color=%23eee';`.

More about [`reshoot`](https://github.com/billykwok/reshoot).

## Installation

```sh
# PNPM
pnpm add -D @reshoot/macro

# NPM
npm install -D @reshoot/macro

# yarn
yarn add -D @reshoot/macro
```

## Setup

```js
// babel.config.js
module.exports = {
  // ...
  plugins: ['babel-plugin-macros'],
};
```

## API

```js
import reshoot from '@reshoot/macro';

const meta = reshoot(
  './image.png',
  { color: '#eee' } // Optional inline config overriding the loader options
);
```

## Support

This library has been continuously used in many of my personal projects, and is regarded as production-ready. In the foreseeable future, I will continuously maintain and support this library.

## Issues and Feedback

Please voice your opinion and report bugs in the [issues](https://github.com/billykwok/reshoot/issues) sections of this GitHub project.

## Contributing

You are more than welcome to add more functionalities, improve documentation, fix bugs, and anything you think is needed. The build step is pretty self-explanatory. Please refer to [`package.json`](https://github.com/billykwok/reshoot/blob/main/package.json).

## License

MIT
