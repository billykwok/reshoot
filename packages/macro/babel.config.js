module.exports = {
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
        useESModules: true,
      },
    ],
    'dynamic-import-node',
  ],
};
