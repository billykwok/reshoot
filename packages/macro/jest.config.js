module.exports = {
  displayName: '@reshoot/macro',
  testEnvironment: 'node',
  snapshotResolver: '<rootDir>/test/snapshotResolver.js',
  coverageProvider: 'v8',
  verbose: true
};
