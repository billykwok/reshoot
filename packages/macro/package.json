{
  "name": "@reshoot/macro",
  "description": "Babel macro for @reshoot/loader",
  "version": "1.1.10",
  "type": "commonjs",
  "exports": {
    "require": "./lib/index.js"
  },
  "main": "./lib/index.js",
  "types": "./types/index.d.ts",
  "files": [
    "lib",
    "types"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "author": "Billy Kwok <opensource@billykwok.me>",
  "license": "MIT",
  "keywords": [
    "reshoot",
    "srcset",
    "image",
    "img",
    "responsive",
    "babel",
    "babel-macros",
    "babel-plugin-macros"
  ],
  "repository": "https://github.com/billykwok/reshoot",
  "bugs": {
    "url": "https://github.com/billykwok/reshoot/issues"
  },
  "homepage": "https://github.com/billykwok/reshoot/tree/master/packages/macro",
  "scripts": {
    "clean": "rm -rf lib",
    "build": "yarn clean && babel src --root-mode=upward --out-dir lib --extensions .ts && tsc --project tsconfig.json",
    "lint": "eslint ./src/**"
  },
  "dependencies": {
    "@babel/types": "^7.11.0"
  },
  "peerDependencies": {
    "@babel/core": "^7.11.1",
    "@reshoot/loader": "*",
    "babel-plugin-macros": "^2.8.0"
  }
}
