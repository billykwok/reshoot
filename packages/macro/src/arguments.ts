import { NodePath } from '@babel/core';
import {
  isExpression,
  isCallExpression,
  isObjectExpression
} from '@babel/types';
import { MacroError } from 'babel-plugin-macros';
import util from 'util';

function inspect(obj: any) {
  return util.inspect(obj, false, null, true);
}

export default function extractArgumentPaths(referencePath: NodePath) {
  if (!isCallExpression(referencePath.parentPath)) {
    throw new MacroError('Please use it as a function.');
  }

  const argumentPaths = referencePath.parentPath.get('arguments');
  if (!Array.isArray(argumentPaths)) {
    throw new Error('Expect an array of arguments');
  }
  const numberOfArguments = argumentPaths.length;
  if (numberOfArguments < 1 || numberOfArguments > 2) {
    throw new MacroError(
      'Invalid number of arguments: ' + numberOfArguments + '. Expected: 1 to 2'
    );
  }

  if (!isExpression(argumentPaths[0].node)) {
    throw new MacroError(
      `The first argument must be an expression, got ${inspect(
        argumentPaths[0]
      )}.`
    );
  }
  const pathEvaluation = argumentPaths[0].evaluate();

  if (!pathEvaluation.confident || typeof pathEvaluation.value !== 'string') {
    throw new MacroError(
      `Failed to evaluate argument ${inspect(argumentPaths[0])}.`
    );
  }
  console.info(`Received argument ${pathEvaluation.value}.`);

  if (numberOfArguments === 1) {
    return [pathEvaluation.value, {}];
  }

  if (!isObjectExpression(argumentPaths[1].node)) {
    throw new MacroError('The second argument must be an object expression.');
  }

  const optionsEvaluation = argumentPaths[1].evaluate();

  if (!optionsEvaluation.confident) {
    throw new MacroError('Failed to evaluate options.');
  }

  return [pathEvaluation.value, optionsEvaluation.value];
}
