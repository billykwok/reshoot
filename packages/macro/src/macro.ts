import { createMacro } from 'babel-plugin-macros';

import handle from './handle';

export default createMacro(handle) as (
  path: string,
  options: Readonly<
    Partial<{
      name: string;
      outputPath: string | ((filename: string) => string);
      publicPath: string | ((filename: string) => string);
      shape: {
        mime: string | boolean;
        src: string | boolean;
        width: string | boolean;
        height: string | boolean;
        srcSet: string | boolean;
        aspectRatio: string | boolean;
        placeholder: string | boolean;
        color: string | boolean;
      };
      srcSet: number[];
      quality: number;
      background: string;
      color: string;
      enforceFormat: string | false;
      placeholder: { size: number; trimDataUrl: boolean };
      aspectRatio: {
        type: 'widthByHeight' | 'heightByWidth';
        format: 'ratio' | 'percent';
        decimal: number;
      };
      disable: boolean;
      cache: boolean;
      emitFile: boolean;
      esModule: boolean;
      [key: string]: unknown;
    }>
  >
) => Readonly<
  Partial<{
    src: string;
    width: number;
    height: number;
    aspectRatio: number;
    srcSet: string;
    alt: string;
    color: string;
    placeholder: string;
    [key: string]: unknown;
  }>
>;
