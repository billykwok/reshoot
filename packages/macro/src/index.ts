import macro from './macro';

export type { ImageData } from './macro';

export default macro;
