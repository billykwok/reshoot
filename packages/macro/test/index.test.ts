import { describe, expect, test } from '@jest/globals';
import { format } from 'prettier';
import { transformAsync, type TransformOptions } from '@babel/core';

expect.addSnapshotSerializer({
  serialize: (value: string) => format(value, { parser: 'babel' }),
  test: (value: string) => typeof value === 'string',
});

const BABEL_TRANSFORM_OPTIONS: TransformOptions = {
  rootMode: 'root',
  babelrc: false,
  presets: [
    [
      '@babel/preset-env',
      {
        targets: { node: 'current', esmodules: false },
        exclude: ['transform-async-to-generator', 'transform-regenerator'],
        modules: false,
      },
    ],
  ],
  plugins: ['babel-plugin-macros'],
};

describe('Babel with @reshoot/macro', () => {
  test('should export the same function from `index.ts` and `macro.ts`', async () => {
    const [index, macro] = await Promise.all([
      import('../src'),
      import('../src/macro'),
    ]);
    expect(index).toStrictEqual(macro);
  });

  test('should do nothing when macro is unused', async () => {
    const result = await transformAsync(
      `
        import { createElement } from 'react';
        import imageMeta from './macro/src/macro';

        export default function ExampleImg(props) {
          return createElement(img, props);
        };
      `,
      BABEL_TRANSFORM_OPTIONS
    );
    expect(result.code).toMatchSnapshot();
  });

  test('should correctly transform', async () => {
    const result = await transformAsync(
      `
        import Img from '@reshoot/react';
        import { createElement } from 'react';
        import imageMeta from './macro/src/macro';

        const meta = imageMeta('../asset/image.png?color=#f4f4f4');
        export default function ExampleImg(props) {
          return createElement(Img, { meta, ...props });
        };
      `,
      BABEL_TRANSFORM_OPTIONS
    );
    expect(result.code).toMatchSnapshot();
  });

  test('should correctly transform with two identical sources', async () => {
    const result = await transformAsync(
      `
        import Img from '@reshoot/react';
        import { createElement } from 'react';
        import imageMeta from './macro/src/macro';

        const meta = imageMeta('../asset/image.png', { color: '#f4f4f4' });
        export default function ExampleImg(props) {
          return createElement(Img, { meta, ...props });
        };

        export const meta2 = imageMeta('../asset/image.png?color=%23f4f4f4');
      `,
      BABEL_TRANSFORM_OPTIONS
    );
    expect(result.code).toMatchSnapshot();
  });

  test('should correctly transform with options', async () => {
    const result = await transformAsync(
      `
        import Img from '@reshoot/react';
        import { createElement } from 'react';
        import imageMeta from './macro/src/macro';

        const meta = imageMeta('../asset/image.png', { color: '#f4f4f4' });
        export default function ExampleImg(props) {
          return createElement(Img, { meta, ...props });
        };
      `,
      BABEL_TRANSFORM_OPTIONS
    );
    expect(result.code).toMatchSnapshot();
  });

  test('should correctly transform with options and existing search params', async () => {
    const result = await transformAsync(
      `
        import Img from '@reshoot/react';
        import { createElement } from 'react';
        import imageMeta from './macro/src/macro';

        const meta = imageMeta('../asset/image.png?reshoot', { color: '#f4f4f4' });
        export default function ExampleImg(props) {
          return createElement(Img, { meta, ...props });
        };
      `,
      BABEL_TRANSFORM_OPTIONS
    );
    expect(result.code).toMatchSnapshot();
  });

  test('should throw if there are no arguments', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta();
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if there are more than two arguments', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta('../asset/image.png', { color: '#f4f4f4' }, 1);
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if not used as a function', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta;
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if path cannot be evaluated', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta('../asset/image.png' + new Date().toString());
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if path is not evaluated into a string', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta(1 + 2 + 3);
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if path is a HTTP url', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta('https://example.com/image.png');
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if options cannot be evaluated', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta('../asset/image.png', { color: new Date().toString() });
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if options is not evaluated into an object', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta('../asset/image.png', 1);
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });

  test('should throw if there is unknown options', async () => {
    expect.assertions(1);
    try {
      await transformAsync(
        `
          import Img from '@reshoot/react';
          import { createElement } from 'react';
          import imageMeta from './macro/src/macro';

          const meta = imageMeta('../asset/image.png', { foo: 'bar' });;
          export default function ExampleImg(props) {
            return createElement(Img, { meta, ...props });
          };
        `,
        BABEL_TRANSFORM_OPTIONS
      );
    } catch (e) {
      // eslint-disable-next-line jest/no-conditional-expect
      expect(e).toMatchSnapshot();
    }
  });
});
