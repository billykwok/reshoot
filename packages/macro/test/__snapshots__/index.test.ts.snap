// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Babel with @reshoot/macro should correctly transform 1`] = `
import __0____asset_image_png_color__f4f4f4__ from "../asset/image.png?color=#f4f4f4";
import Img from "@reshoot/react";
import { createElement } from "react";
const meta = __0____asset_image_png_color__f4f4f4__;
export default function ExampleImg(props) {
  return createElement(Img, { meta, ...props });
}

`;

exports[`Babel with @reshoot/macro should correctly transform with options 1`] = `
import __0____asset_image_png_color__23f4f4f4__ from "../asset/image.png?color=%23f4f4f4";
import Img from "@reshoot/react";
import { createElement } from "react";
const meta = __0____asset_image_png_color__23f4f4f4__;
export default function ExampleImg(props) {
  return createElement(Img, { meta, ...props });
}

`;

exports[`Babel with @reshoot/macro should correctly transform with options and existing search params 1`] = `
import __0____asset_image_png_reshoot_color__23f4f4f4__ from "../asset/image.png?reshoot&color=%23f4f4f4";
import Img from "@reshoot/react";
import { createElement } from "react";
const meta = __0____asset_image_png_reshoot_color__23f4f4f4__;
export default function ExampleImg(props) {
  return createElement(Img, { meta, ...props });
}

`;

exports[`Babel with @reshoot/macro should correctly transform with two identical sources 1`] = `
import __0____asset_image_png_color__23f4f4f4__ from "../asset/image.png?color=%23f4f4f4";
import Img from "@reshoot/react";
import { createElement } from "react";
const meta = __0____asset_image_png_color__23f4f4f4__;
export default function ExampleImg(props) {
  return createElement(Img, { meta, ...props });
}
export const meta2 = __0____asset_image_png_color__23f4f4f4__;

`;

exports[`Babel with @reshoot/macro should do nothing when macro is unused 1`] = `
import { createElement } from "react";
export default function ExampleImg(props) {
  return createElement(img, props);
}

`;

exports[`Babel with @reshoot/macro should throw if not used as a function 1`] = `[MacroError: unknown: Please use it as a function]`;

exports[`Babel with @reshoot/macro should throw if options cannot be evaluated 1`] = `
[MacroError: unknown: Failed to evaluate Reshoot options Node {
  type: [32m'ObjectExpression'[39m,
  start: [33m203[39m,
  end: [33m235[39m,
  loc: SourceLocation {
    start: Position { line: [33m6[39m, column: [33m55[39m, index: [33m203[39m },
    end: Position { line: [33m6[39m, column: [33m87[39m, index: [33m235[39m },
    filename: [90mundefined[39m,
    identifierName: [90mundefined[39m
  },
  properties: [
    Node {
      type: [32m'ObjectProperty'[39m,
      start: [33m205[39m,
      end: [33m233[39m,
      loc: SourceLocation {
        start: Position { line: [33m6[39m, column: [33m57[39m, index: [33m205[39m },
        end: Position { line: [33m6[39m, column: [33m85[39m, index: [33m233[39m },
        filename: [90mundefined[39m,
        identifierName: [90mundefined[39m
      },
      method: [33mfalse[39m,
      key: Node {
        type: [32m'Identifier'[39m,
        start: [33m205[39m,
        end: [33m210[39m,
        loc: SourceLocation {
          start: Position { line: [33m6[39m, column: [33m57[39m, index: [33m205[39m },
          end: Position { line: [33m6[39m, column: [33m62[39m, index: [33m210[39m },
          filename: [90mundefined[39m,
          identifierName: [32m'color'[39m
        },
        name: [32m'color'[39m,
        leadingComments: [90mundefined[39m,
        innerComments: [90mundefined[39m,
        trailingComments: [90mundefined[39m
      },
      computed: [33mfalse[39m,
      shorthand: [33mfalse[39m,
      value: Node {
        type: [32m'CallExpression'[39m,
        start: [33m212[39m,
        end: [33m233[39m,
        loc: SourceLocation {
          start: Position { line: [33m6[39m, column: [33m64[39m, index: [33m212[39m },
          end: Position { line: [33m6[39m, column: [33m85[39m, index: [33m233[39m },
          filename: [90mundefined[39m,
          identifierName: [90mundefined[39m
        },
        callee: Node {
          type: [32m'MemberExpression'[39m,
          start: [33m212[39m,
          end: [33m231[39m,
          loc: SourceLocation {
            start: Position { line: [33m6[39m, column: [33m64[39m, index: [33m212[39m },
            end: Position { line: [33m6[39m, column: [33m83[39m, index: [33m231[39m },
            filename: [90mundefined[39m,
            identifierName: [90mundefined[39m
          },
          object: Node {
            type: [32m'NewExpression'[39m,
            start: [33m212[39m,
            end: [33m222[39m,
            loc: SourceLocation {
              start: Position { line: [33m6[39m, column: [33m64[39m, index: [33m212[39m },
              end: Position { line: [33m6[39m, column: [33m74[39m, index: [33m222[39m },
              filename: [90mundefined[39m,
              identifierName: [90mundefined[39m
            },
            callee: Node {
              type: [32m'Identifier'[39m,
              start: [33m216[39m,
              end: [33m220[39m,
              loc: SourceLocation {
                start: Position { line: [33m6[39m, column: [33m68[39m, index: [33m216[39m },
                end: Position { line: [33m6[39m, column: [33m72[39m, index: [33m220[39m },
                filename: [90mundefined[39m,
                identifierName: [32m'Date'[39m
              },
              name: [32m'Date'[39m,
              leadingComments: [90mundefined[39m,
              innerComments: [90mundefined[39m,
              trailingComments: [90mundefined[39m
            },
            arguments: [],
            leadingComments: [90mundefined[39m,
            innerComments: [90mundefined[39m,
            trailingComments: [90mundefined[39m
          },
          computed: [33mfalse[39m,
          property: Node {
            type: [32m'Identifier'[39m,
            start: [33m223[39m,
            end: [33m231[39m,
            loc: SourceLocation {
              start: Position { line: [33m6[39m, column: [33m75[39m, index: [33m223[39m },
              end: Position { line: [33m6[39m, column: [33m83[39m, index: [33m231[39m },
              filename: [90mundefined[39m,
              identifierName: [32m'toString'[39m
            },
            name: [32m'toString'[39m,
            leadingComments: [90mundefined[39m,
            innerComments: [90mundefined[39m,
            trailingComments: [90mundefined[39m
          },
          leadingComments: [90mundefined[39m,
          innerComments: [90mundefined[39m,
          trailingComments: [90mundefined[39m
        },
        arguments: [],
        leadingComments: [90mundefined[39m,
        innerComments: [90mundefined[39m,
        trailingComments: [90mundefined[39m
      },
      leadingComments: [90mundefined[39m,
      innerComments: [90mundefined[39m,
      trailingComments: [90mundefined[39m
    }
  ],
  leadingComments: [90mundefined[39m,
  innerComments: [90mundefined[39m,
  trailingComments: [90mundefined[39m
}]
`;

exports[`Babel with @reshoot/macro should throw if options is not evaluated into an object 1`] = `[MacroError: unknown: Reshoot options must be evaluated into a string, but got [33m1[39m]`;

exports[`Babel with @reshoot/macro should throw if path cannot be evaluated 1`] = `
[MacroError: unknown: Failed to evaluate Reshoot path Node {
  type: [32m'BinaryExpression'[39m,
  start: [33m181[39m,
  end: [33m225[39m,
  loc: SourceLocation {
    start: Position { line: [33m6[39m, column: [33m33[39m, index: [33m181[39m },
    end: Position { line: [33m6[39m, column: [33m77[39m, index: [33m225[39m },
    filename: [90mundefined[39m,
    identifierName: [90mundefined[39m
  },
  left: Node {
    type: [32m'StringLiteral'[39m,
    start: [33m181[39m,
    end: [33m201[39m,
    loc: SourceLocation {
      start: Position { line: [33m6[39m, column: [33m33[39m, index: [33m181[39m },
      end: Position { line: [33m6[39m, column: [33m53[39m, index: [33m201[39m },
      filename: [90mundefined[39m,
      identifierName: [90mundefined[39m
    },
    extra: { rawValue: [32m'../asset/image.png'[39m, raw: [32m"'../asset/image.png'"[39m },
    value: [32m'../asset/image.png'[39m,
    leadingComments: [90mundefined[39m,
    innerComments: [90mundefined[39m,
    trailingComments: [90mundefined[39m
  },
  operator: [32m'+'[39m,
  right: Node {
    type: [32m'CallExpression'[39m,
    start: [33m204[39m,
    end: [33m225[39m,
    loc: SourceLocation {
      start: Position { line: [33m6[39m, column: [33m56[39m, index: [33m204[39m },
      end: Position { line: [33m6[39m, column: [33m77[39m, index: [33m225[39m },
      filename: [90mundefined[39m,
      identifierName: [90mundefined[39m
    },
    callee: Node {
      type: [32m'MemberExpression'[39m,
      start: [33m204[39m,
      end: [33m223[39m,
      loc: SourceLocation {
        start: Position { line: [33m6[39m, column: [33m56[39m, index: [33m204[39m },
        end: Position { line: [33m6[39m, column: [33m75[39m, index: [33m223[39m },
        filename: [90mundefined[39m,
        identifierName: [90mundefined[39m
      },
      object: Node {
        type: [32m'NewExpression'[39m,
        start: [33m204[39m,
        end: [33m214[39m,
        loc: SourceLocation {
          start: Position { line: [33m6[39m, column: [33m56[39m, index: [33m204[39m },
          end: Position { line: [33m6[39m, column: [33m66[39m, index: [33m214[39m },
          filename: [90mundefined[39m,
          identifierName: [90mundefined[39m
        },
        callee: Node {
          type: [32m'Identifier'[39m,
          start: [33m208[39m,
          end: [33m212[39m,
          loc: SourceLocation {
            start: Position { line: [33m6[39m, column: [33m60[39m, index: [33m208[39m },
            end: Position { line: [33m6[39m, column: [33m64[39m, index: [33m212[39m },
            filename: [90mundefined[39m,
            identifierName: [32m'Date'[39m
          },
          name: [32m'Date'[39m,
          leadingComments: [90mundefined[39m,
          innerComments: [90mundefined[39m,
          trailingComments: [90mundefined[39m
        },
        arguments: [],
        leadingComments: [90mundefined[39m,
        innerComments: [90mundefined[39m,
        trailingComments: [90mundefined[39m
      },
      computed: [33mfalse[39m,
      property: Node {
        type: [32m'Identifier'[39m,
        start: [33m215[39m,
        end: [33m223[39m,
        loc: SourceLocation {
          start: Position { line: [33m6[39m, column: [33m67[39m, index: [33m215[39m },
          end: Position { line: [33m6[39m, column: [33m75[39m, index: [33m223[39m },
          filename: [90mundefined[39m,
          identifierName: [32m'toString'[39m
        },
        name: [32m'toString'[39m,
        leadingComments: [90mundefined[39m,
        innerComments: [90mundefined[39m,
        trailingComments: [90mundefined[39m
      },
      leadingComments: [90mundefined[39m,
      innerComments: [90mundefined[39m,
      trailingComments: [90mundefined[39m
    },
    arguments: [],
    leadingComments: [90mundefined[39m,
    innerComments: [90mundefined[39m,
    trailingComments: [90mundefined[39m
  },
  leadingComments: [90mundefined[39m,
  innerComments: [90mundefined[39m,
  trailingComments: [90mundefined[39m
}]
`;

exports[`Babel with @reshoot/macro should throw if path is a HTTP url 1`] = `[MacroError: unknown: Only local path is supported, but got https://example.com/image.png]`;

exports[`Babel with @reshoot/macro should throw if path is not evaluated into a string 1`] = `[MacroError: unknown: Reshoot path must be evaluated into a string, but got [33m6[39m]`;

exports[`Babel with @reshoot/macro should throw if there are more than two arguments 1`] = `[MacroError: unknown: Expect 1 - 2 argument(s), but got 3]`;

exports[`Babel with @reshoot/macro should throw if there are no arguments 1`] = `[MacroError: unknown: Expect 1 - 2 argument(s), but got 0]`;

exports[`Babel with @reshoot/macro should throw if there is unknown options 1`] = `[MacroError: unknown: Unknown options foo]`;
